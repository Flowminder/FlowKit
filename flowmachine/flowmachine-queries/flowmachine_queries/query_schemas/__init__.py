# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
from .aggregate_network_objects import AggregateNetworkObjectsSchema
from .dfs_metric_total_amount import DFSTotalMetricAmountSchema
from .dummy_query import DummyQuerySchema
from .flows import FlowsSchema
from .geography import GeographySchema
from .joined_spatial_aggregate import JoinedSpatialAggregateSchema
from .location_event_counts import LocationEventCountsSchema
from .location_introversion import LocationIntroversionSchema
from .meaningful_locations import (
    MeaningfulLocationsAggregateSchema,
    MeaningfulLocationsBetweenLabelODMatrixSchema,
    MeaningfulLocationsBetweenDatesODMatrixSchema,
)
from .spatial_aggregate import SpatialAggregateSchema
from .total_network_objects import TotalNetworkObjectsSchema
from .unique_subscriber_counts import UniqueSubscriberCountsSchema

schemas = {
    "dummy_query": DummyQuerySchema,
    "flows": FlowsSchema,
    "meaningful_locations_aggregate": MeaningfulLocationsAggregateSchema,
    "meaningful_locations_between_label_od_matrix": MeaningfulLocationsBetweenLabelODMatrixSchema,
    "meaningful_locations_between_dates_od_matrix": MeaningfulLocationsBetweenDatesODMatrixSchema,
    "geography": GeographySchema,
    "location_event_counts": LocationEventCountsSchema,
    "unique_subscriber_counts": UniqueSubscriberCountsSchema,
    "location_introversion": LocationIntroversionSchema,
    "total_network_objects": TotalNetworkObjectsSchema,
    "aggregate_network_objects": AggregateNetworkObjectsSchema,
    "dfs_metric_total_amount": DFSTotalMetricAmountSchema,
    "spatial_aggregate": SpatialAggregateSchema,
    "joined_spatial_aggregate": JoinedSpatialAggregateSchema,
}
