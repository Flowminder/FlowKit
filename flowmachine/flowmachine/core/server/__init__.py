# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import logging
import sys

import structlog


# Logger for all queries run or accessed
query_run_log = logging.getLogger("flowmachine").getChild("query_run_log")
ch = logging.StreamHandler(sys.stdout)
ch.setLevel(logging.INFO)
query_run_log.addHandler(ch)
query_run_log = structlog.wrap_logger(query_run_log)
