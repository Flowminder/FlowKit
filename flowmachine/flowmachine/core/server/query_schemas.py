import json

from hashlib import md5
from marshmallow import Schema, fields, post_load
from marshmallow.validate import OneOf, Length
from marshmallow_oneofschema import OneOfSchema


#
# Flowmachine query schemas
#


class DailyLocationSchema(Schema):
    date = fields.Date(required=True)
    method = fields.String(required=True, validate=OneOf(["last", "most-common"]))
    aggregation_unit = fields.String(
        required=True, validate=OneOf(["admin0", "admin1", "admin2", "admin3"])
    )
    subscriber_subset = fields.String(
        required=False, allow_none=True, validate=OneOf([None])
    )

    @post_load
    def make_query_object(self, params):
        return DailyLocationExposed(**params)


class InputToModalLocationSchema(OneOfSchema):
    type_field = "query_kind"
    type_schemas = {"daily_location": DailyLocationSchema}

    def get_obj_type(self, obj):
        if isinstance(obj, DailyLocationExposed):
            return "daily_location"
        else:
            raise Exception("Unknown object type: {obj.__class__.__name__}")


class ModalLocationSchema(Schema):
    locations = fields.Nested(
        InputToModalLocationSchema, many=True, validate=Length(min=1)
    )
    aggregation_unit = fields.String(
        validate=OneOf(["admin0", "admin1", "admin2", "admin3"])
    )
    subscriber_subset = fields.String(allow_none=True, validate=OneOf([None]))

    @post_load
    def make_query_object(self, data):
        return ModalLocationExposed(**data)


class FlowmachineQuerySchema(OneOfSchema):
    type_field = "query_kind"
    type_schemas = {
        "daily_location": DailyLocationSchema,
        "modal_location": ModalLocationSchema,
    }

    def get_obj_type(self, obj):
        if isinstance(obj, DailyLocationExposed):
            return "daily_location"
        elif isinstance(obj, ModalLocationExposed):
            return "modal_location"
        else:
            raise Exception("Unknown object type: {obj.__class__.__name__}")


#
# Flowmachine query objects
#


class BaseExposedQuery:
    """
    Base class for flowmachine queries.
    """

    @property
    def query_id(self):
        return md5(json.dumps(self.query_params, sort_keys=True).encode()).hexdigest()

    @property
    def query_params(self):
        marshmallow_schema = self.__schema__()
        return marshmallow_schema.dump(self)


class DailyLocationExposed(BaseExposedQuery):

    __schema__ = DailyLocationSchema

    def __init__(self, date, *, method, aggregation_unit, subscriber_subset=None):
        self.date = date
        self.method = method
        self.aggregation_unit = aggregation_unit
        self.subscriber_subset = subscriber_subset


class ModalLocationExposed(BaseExposedQuery):

    __schema__ = ModalLocationSchema

    def __init__(self, locations, *, aggregation_unit, subscriber_subset=None):
        self.locations = locations
        self.aggregation_unit = aggregation_unit
        self.subscriber_subset = subscriber_subset
