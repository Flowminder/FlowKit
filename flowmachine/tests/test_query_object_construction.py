# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import json
import pytest
from marshmallow import ValidationError

from flowmachine.core.server.query_schemas import FlowmachineQuerySchema


def test_construct_query(diff_reporter):
    """
    Test that query objects constructed by construct_query_object() have the expected query ids.
    """
    query_specs_to_test = [
        {
            "query_kind": "spatial_aggregate",
            "locations": {
                "query_kind": "daily_location",
                "date": "2016-01-01",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": None,
                "sampling": {
                    "sampling_method": "bernoulli",
                    "size": 10,
                    "seed": 0.5,
                    "fraction": None,
                    "estimate_count": False,
                },
            },
        },
        {
            "query_kind": "spatial_aggregate",
            "locations": {
                "query_kind": "daily_location",
                "date": "2016-01-01",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": None,
            },
        },
        {
            "query_kind": "spatial_aggregate",
            "locations": {
                "query_kind": "daily_location",
                "date": "2016-01-01",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": None,
                "sampling": None,
            },
        },
        {
            "query_kind": "location_event_counts",
            "start_date": "2016-01-01",
            "end_date": "2016-01-02",
            "interval": "day",
            "aggregation_unit": "admin3",
            "direction": "both",
            "event_types": None,
            "subscriber_subset": None,
        },
        {
            "query_kind": "spatial_aggregate",
            "locations": {
                "query_kind": "modal_location",
                "locations": (
                    {
                        "query_kind": "daily_location",
                        "date": "2016-01-01",
                        "aggregation_unit": "admin3",
                        "method": "last",
                        "subscriber_subset": None,
                    },
                    {
                        "query_kind": "daily_location",
                        "date": "2016-01-02",
                        "aggregation_unit": "admin3",
                        "method": "last",
                        "subscriber_subset": None,
                    },
                ),
            },
        },
        {"query_kind": "geography", "aggregation_unit": "admin3"},
        {
            "query_kind": "meaningful_locations_aggregate",
            "aggregation_unit": "admin1",
            "start_date": "2016-01-01",
            "end_date": "2016-01-02",
            "label": "unknown",
            "labels": {
                "evening": {
                    "type": "Polygon",
                    "coordinates": [
                        [[1e-06, -0.5], [1e-06, -1.1], [1.1, -1.1], [1.1, -0.5]]
                    ],
                },
                "day": {
                    "type": "Polygon",
                    "coordinates": [
                        [[-1.1, -0.5], [-1.1, 0.5], [-1e-06, 0.5], [0, -0.5]]
                    ],
                },
            },
            "tower_hour_of_day_scores": [
                -1,
                -1,
                -1,
                -1,
                -1,
                -1,
                -1,
                0,
                0,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                0,
                0,
                0,
                0,
                -1,
                -1,
                -1,
            ],
            "tower_day_of_week_scores": {
                "monday": 1,
                "tuesday": 1,
                "wednesday": 1,
                "thursday": 0,
                "friday": -1,
                "saturday": -1,
                "sunday": -1,
            },
            "tower_cluster_radius": 1.0,
            "tower_cluster_call_threshold": 0,
            "subscriber_subset": None,
        },
        {
            "query_kind": "meaningful_locations_between_label_od_matrix",
            "aggregation_unit": "admin1",
            "start_date": "2016-01-01",
            "end_date": "2016-01-02",
            "label_a": "day",
            "label_b": "evening",
            "labels": {
                "day": {
                    "type": "Polygon",
                    "coordinates": [
                        [[-1.1, -0.5], [-1.1, 0.5], [-1e-06, 0.5], [0, -0.5]]
                    ],
                },
                "evening": {
                    "type": "Polygon",
                    "coordinates": [
                        [[1e-06, -0.5], [1e-06, -1.1], [1.1, -1.1], [1.1, -0.5]]
                    ],
                },
            },
            "tower_hour_of_day_scores": [
                -1,
                -1,
                -1,
                -1,
                -1,
                -1,
                -1,
                0,
                0,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                0,
                0,
                0,
                0,
                -1,
                -1,
                -1,
            ],
            "tower_day_of_week_scores": {
                "monday": 1,
                "tuesday": 1,
                "wednesday": 1,
                "thursday": 0,
                "friday": -1,
                "saturday": -1,
                "sunday": -1,
            },
            "tower_cluster_radius": 1.0,
            "tower_cluster_call_threshold": 0,
            "subscriber_subset": None,
        },
        {
            "query_kind": "meaningful_locations_between_dates_od_matrix",
            "aggregation_unit": "admin1",
            "start_date_a": "2016-01-01",
            "end_date_a": "2016-01-02",
            "start_date_b": "2016-01-01",
            "end_date_b": "2016-01-05",
            "label": "unknown",
            "labels": {
                "day": {
                    "type": "Polygon",
                    "coordinates": [
                        [[-1.1, -0.5], [-1.1, 0.5], [-1e-06, 0.5], [0, -0.5]]
                    ],
                },
                "evening": {
                    "type": "Polygon",
                    "coordinates": [
                        [[1e-06, -0.5], [1e-06, -1.1], [1.1, -1.1], [1.1, -0.5]]
                    ],
                },
            },
            "tower_hour_of_day_scores": [
                -1,
                -1,
                -1,
                -1,
                -1,
                -1,
                -1,
                0,
                0,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                1,
                0,
                0,
                0,
                0,
                -1,
                -1,
                -1,
            ],
            "tower_day_of_week_scores": {
                "monday": 1,
                "tuesday": 1,
                "wednesday": 1,
                "thursday": 0,
                "friday": -1,
                "saturday": -1,
                "sunday": -1,
            },
            "tower_cluster_radius": 1.0,
            "tower_cluster_call_threshold": 2,
            "subscriber_subset": None,
        },
    ]

    def get_query_id_for_query_spec(query_spec):
        return FlowmachineQuerySchema().load(query_spec).query_id

    query_ids_and_specs_as_json_string = json.dumps(
        {get_query_id_for_query_spec(spec): spec for spec in query_specs_to_test},
        indent=2,
    )

    diff_reporter(query_ids_and_specs_as_json_string)


# TODO: we should re-think how we want to test invalid values, now that these are validated using marshmallow
def test_wrong_geography_aggregation_unit_raises_error():
    """
    Test that an invalid aggregation unit in a geography query raises an InvalidGeographyError
    """
    with pytest.raises(
        ValidationError,
        match="aggregation_unit.*Must be one of: admin0, admin1, admin2, admin3.",
    ):
        _ = FlowmachineQuerySchema().load(
            {"query_kind": "geography", "aggregation_unit": "DUMMY_AGGREGATION_UNIT"}
        )


@pytest.mark.parametrize(
    "sampling, message",
    [
        (
            {"sampling_method": "bernoulli", "size": 10, "fraction": 0.2},
            "Missing data for required field.",
        ),
        (
            {"sampling_method": "bernoulli", "size": 10, "fraction": 0.2, "seed": 0.1},
            "Must provide exactly one of 'size' or 'fraction' for a random sample",
        ),
        (
            {"sampling_method": "bernoulli", "seed": 0.1},
            "Must provide exactly one of 'size' or 'fraction' for a random sample",
        ),
        (
            {"sampling_method": "bernoulli", "fraction": 1.2, "seed": 0.1},
            "Must be greater than 0.0 and less than 1.0.",
        ),
        (
            {"sampling_method": "bernoulli", "size": -1, "seed": 0.1},
            "Must be greater than or equal to 1.",
        ),
        (
            {"sampling_method": "random_ids", "size": 10, "seed": 185},
            "Must be greater than or equal to -1.0 and less than or equal to 1.0.",
        ),
    ],
)
def test_invalid_sampling_params_raises_error(sampling, message):
    query_spec = {
        "query_kind": "spatial_aggregate",
        "locations": {
            "query_kind": "daily_location",
            "date": "2016-01-01",
            "aggregation_unit": "admin3",
            "method": "last",
            "sampling": sampling,
        },
    }
    with pytest.raises(ValidationError, match=message) as exc:
        _ = FlowmachineQuerySchema().load(query_spec)
    print(exc)
