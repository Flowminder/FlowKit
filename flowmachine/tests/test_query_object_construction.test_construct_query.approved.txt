{
  "261b85fc9c64253889174ea8151f1d69": {
    "query_kind": "spatial_aggregate",
    "locations": {
      "query_kind": "daily_location",
      "date": "2016-01-01",
      "aggregation_unit": "admin3",
      "method": "last",
      "event_types": [
        "calls",
        "sms"
      ],
      "subscriber_subset": null,
      "sampling": {
        "sampling_method": "bernoulli",
        "size": 10,
        "seed": 0.5,
        "fraction": null,
        "estimate_count": false
      }
    }
  },
  "e142ffb3174add433422c2724a08c02b": {
    "query_kind": "spatial_aggregate",
    "locations": {
      "query_kind": "daily_location",
      "date": "2016-01-01",
      "aggregation_unit": "admin3",
      "method": "last",
      "event_types": null,
      "subscriber_subset": null,
      "sampling": null
    }
  },
  "9ddadf16ed7bb751a3e764ec4ccb48c3": {
    "query_kind": "location_event_counts",
    "start_date": "2016-01-01",
    "end_date": "2016-01-02",
    "interval": "day",
    "aggregation_unit": "admin3",
    "direction": "both",
    "event_types": null,
    "subscriber_subset": null
  },
  "b48d02838163766771eeed8cd8aafd98": {
    "query_kind": "spatial_aggregate",
    "locations": {
      "query_kind": "modal_location",
      "locations": [
        {
          "query_kind": "daily_location",
          "date": "2016-01-01",
          "aggregation_unit": "admin3",
          "method": "last",
          "subscriber_subset": null
        },
        {
          "query_kind": "daily_location",
          "date": "2016-01-02",
          "aggregation_unit": "admin3",
          "method": "last",
          "subscriber_subset": null
        }
      ]
    }
  },
  "6521353e7563ed700dfd2cf90721934b": {
    "query_kind": "geography",
    "aggregation_unit": "admin3"
  },
  "75b0532a747b473f69b45b78fdc29865": {
    "query_kind": "meaningful_locations_aggregate",
    "aggregation_unit": "admin1",
    "start_date": "2016-01-01",
    "end_date": "2016-01-02",
    "label": "unknown",
    "labels": {
      "evening": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              1e-06,
              -0.5
            ],
            [
              1e-06,
              -1.1
            ],
            [
              1.1,
              -1.1
            ],
            [
              1.1,
              -0.5
            ]
          ]
        ]
      },
      "day": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -1.1,
              -0.5
            ],
            [
              -1.1,
              0.5
            ],
            [
              -1e-06,
              0.5
            ],
            [
              0,
              -0.5
            ]
          ]
        ]
      }
    },
    "tower_hour_of_day_scores": [
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      0,
      0,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      0,
      0,
      0,
      0,
      -1,
      -1,
      -1
    ],
    "tower_day_of_week_scores": {
      "monday": 1,
      "tuesday": 1,
      "wednesday": 1,
      "thursday": 0,
      "friday": -1,
      "saturday": -1,
      "sunday": -1
    },
    "tower_cluster_radius": 1.0,
    "tower_cluster_call_threshold": 0,
    "subscriber_subset": null
  },
  "5df56eee4dc96ed961f8eb76583cc6de": {
    "query_kind": "meaningful_locations_between_label_od_matrix",
    "aggregation_unit": "admin1",
    "start_date": "2016-01-01",
    "end_date": "2016-01-02",
    "label_a": "day",
    "label_b": "evening",
    "labels": {
      "day": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -1.1,
              -0.5
            ],
            [
              -1.1,
              0.5
            ],
            [
              -1e-06,
              0.5
            ],
            [
              0,
              -0.5
            ]
          ]
        ]
      },
      "evening": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              1e-06,
              -0.5
            ],
            [
              1e-06,
              -1.1
            ],
            [
              1.1,
              -1.1
            ],
            [
              1.1,
              -0.5
            ]
          ]
        ]
      }
    },
    "tower_hour_of_day_scores": [
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      0,
      0,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      0,
      0,
      0,
      0,
      -1,
      -1,
      -1
    ],
    "tower_day_of_week_scores": {
      "monday": 1,
      "tuesday": 1,
      "wednesday": 1,
      "thursday": 0,
      "friday": -1,
      "saturday": -1,
      "sunday": -1
    },
    "tower_cluster_radius": 1.0,
    "tower_cluster_call_threshold": 0,
    "event_types": null,
    "subscriber_subset": null
  },
  "00bee92b22b1ceb98244c5700a079656": {
    "query_kind": "meaningful_locations_between_dates_od_matrix",
    "aggregation_unit": "admin1",
    "start_date_a": "2016-01-01",
    "end_date_a": "2016-01-02",
    "start_date_b": "2016-01-01",
    "end_date_b": "2016-01-05",
    "label": "unknown",
    "labels": {
      "day": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -1.1,
              -0.5
            ],
            [
              -1.1,
              0.5
            ],
            [
              -1e-06,
              0.5
            ],
            [
              0,
              -0.5
            ]
          ]
        ]
      },
      "evening": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              1e-06,
              -0.5
            ],
            [
              1e-06,
              -1.1
            ],
            [
              1.1,
              -1.1
            ],
            [
              1.1,
              -0.5
            ]
          ]
        ]
      }
    },
    "tower_hour_of_day_scores": [
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      0,
      0,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      0,
      0,
      0,
      0,
      -1,
      -1,
      -1
    ],
    "tower_day_of_week_scores": {
      "monday": 1,
      "tuesday": 1,
      "wednesday": 1,
      "thursday": 0,
      "friday": -1,
      "saturday": -1,
      "sunday": -1
    },
    "tower_cluster_radius": 1.0,
    "tower_cluster_call_threshold": 2,
    "event_types": [
      "calls",
      "sms"
    ],
    "subscriber_subset": null
  },
  "c3dc95da89aeb0908b9398918dab6f26": {
    "query_kind": "flows",
    "from_location": {
      "query_kind": "daily_location",
      "date": "2016-01-01",
      "aggregation_unit": "admin3",
      "method": "last"
    },
    "to_location": {
      "query_kind": "unique_locations",
      "start_date": "2016-01-01",
      "end_date": "2016-01-04",
      "aggregation_unit": "admin3"
    },
    "join_type": "left outer"
  },
  "95fbc18554e15733df47bfd5cbaa3f87": {
    "query_kind": "flows",
    "from_location": {
      "query_kind": "majority_location",
      "subscriber_location_weights": {
        "query_kind": "location_visits",
        "locations": [
          {
            "query_kind": "daily_location",
            "date": "2016-01-01",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          },
          {
            "query_kind": "daily_location",
            "date": "2016-01-02",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          }
        ]
      }
    },
    "to_location": {
      "query_kind": "majority_location",
      "subscriber_location_weights": {
        "query_kind": "location_visits",
        "locations": [
          {
            "query_kind": "daily_location",
            "date": "2016-01-04",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          },
          {
            "query_kind": "daily_location",
            "date": "2016-01-05",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          }
        ]
      }
    }
  },
  "920dfdf5568d75921c4173da8bccc6ef": {
    "query_kind": "labelled_spatial_aggregate",
    "locations": {
      "query_kind": "coalesced_location",
      "preferred_location": {
        "query_kind": "majority_location",
        "subscriber_location_weights": {
          "query_kind": "location_visits",
          "locations": [
            {
              "query_kind": "daily_location",
              "date": "2016-01-03",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            },
            {
              "query_kind": "daily_location",
              "date": "2016-01-04",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            }
          ]
        },
        "minimum_total_weight": 2,
        "include_unlocatable": true
      },
      "fallback_location": {
        "query_kind": "majority_location",
        "subscriber_location_weights": {
          "query_kind": "location_visits",
          "locations": [
            {
              "query_kind": "daily_location",
              "date": "2016-01-01",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            },
            {
              "query_kind": "daily_location",
              "date": "2016-01-02",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            }
          ]
        },
        "minimum_total_weight": 2,
        "include_unlocatable": false
      },
      "subscriber_location_weights": {
        "query_kind": "location_visits",
        "locations": [
          {
            "query_kind": "daily_location",
            "date": "2016-01-03",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          },
          {
            "query_kind": "daily_location",
            "date": "2016-01-04",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          }
        ]
      },
      "weight_threshold": 2
    },
    "labels": {
      "query_kind": "mobility_classification",
      "locations": [
        {
          "query_kind": "coalesced_location",
          "preferred_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-03",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-04",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": true
          },
          "fallback_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-01",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-02",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": false
          },
          "subscriber_location_weights": {
            "query_kind": "location_visits",
            "locations": [
              {
                "query_kind": "daily_location",
                "date": "2016-01-03",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              },
              {
                "query_kind": "daily_location",
                "date": "2016-01-04",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              }
            ]
          },
          "weight_threshold": 2
        },
        {
          "query_kind": "coalesced_location",
          "preferred_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-05",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-06",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": true
          },
          "fallback_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-03",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-04",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": false
          },
          "subscriber_location_weights": {
            "query_kind": "location_visits",
            "locations": [
              {
                "query_kind": "daily_location",
                "date": "2016-01-05",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              },
              {
                "query_kind": "daily_location",
                "date": "2016-01-06",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              }
            ]
          },
          "weight_threshold": 2
        }
      ],
      "stay_length_threshold": 2
    }
  },
  "c915d87b66df83904634990e2f78da9e": {
    "query_kind": "labelled_flows",
    "from_location": {
      "query_kind": "coalesced_location",
      "preferred_location": {
        "query_kind": "majority_location",
        "subscriber_location_weights": {
          "query_kind": "location_visits",
          "locations": [
            {
              "query_kind": "daily_location",
              "date": "2016-01-03",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            },
            {
              "query_kind": "daily_location",
              "date": "2016-01-04",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            }
          ]
        },
        "minimum_total_weight": 2,
        "include_unlocatable": true
      },
      "fallback_location": {
        "query_kind": "majority_location",
        "subscriber_location_weights": {
          "query_kind": "location_visits",
          "locations": [
            {
              "query_kind": "daily_location",
              "date": "2016-01-01",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            },
            {
              "query_kind": "daily_location",
              "date": "2016-01-02",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            }
          ]
        },
        "minimum_total_weight": 2,
        "include_unlocatable": false
      },
      "subscriber_location_weights": {
        "query_kind": "location_visits",
        "locations": [
          {
            "query_kind": "daily_location",
            "date": "2016-01-03",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          },
          {
            "query_kind": "daily_location",
            "date": "2016-01-04",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          }
        ]
      },
      "weight_threshold": 2
    },
    "to_location": {
      "query_kind": "coalesced_location",
      "preferred_location": {
        "query_kind": "majority_location",
        "subscriber_location_weights": {
          "query_kind": "location_visits",
          "locations": [
            {
              "query_kind": "daily_location",
              "date": "2016-01-05",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            },
            {
              "query_kind": "daily_location",
              "date": "2016-01-06",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            }
          ]
        },
        "minimum_total_weight": 2,
        "include_unlocatable": true
      },
      "fallback_location": {
        "query_kind": "majority_location",
        "subscriber_location_weights": {
          "query_kind": "location_visits",
          "locations": [
            {
              "query_kind": "daily_location",
              "date": "2016-01-03",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            },
            {
              "query_kind": "daily_location",
              "date": "2016-01-04",
              "aggregation_unit": "admin3",
              "method": "last",
              "subscriber_subset": null
            }
          ]
        },
        "minimum_total_weight": 2,
        "include_unlocatable": false
      },
      "subscriber_location_weights": {
        "query_kind": "location_visits",
        "locations": [
          {
            "query_kind": "daily_location",
            "date": "2016-01-05",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          },
          {
            "query_kind": "daily_location",
            "date": "2016-01-06",
            "aggregation_unit": "admin3",
            "method": "last",
            "subscriber_subset": null
          }
        ]
      },
      "weight_threshold": 2
    },
    "labels": {
      "query_kind": "mobility_classification",
      "locations": [
        {
          "query_kind": "coalesced_location",
          "preferred_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-03",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-04",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": true
          },
          "fallback_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-01",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-02",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": false
          },
          "subscriber_location_weights": {
            "query_kind": "location_visits",
            "locations": [
              {
                "query_kind": "daily_location",
                "date": "2016-01-03",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              },
              {
                "query_kind": "daily_location",
                "date": "2016-01-04",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              }
            ]
          },
          "weight_threshold": 2
        },
        {
          "query_kind": "coalesced_location",
          "preferred_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-05",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-06",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": true
          },
          "fallback_location": {
            "query_kind": "majority_location",
            "subscriber_location_weights": {
              "query_kind": "location_visits",
              "locations": [
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-03",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                },
                {
                  "query_kind": "daily_location",
                  "date": "2016-01-04",
                  "aggregation_unit": "admin3",
                  "method": "last",
                  "subscriber_subset": null
                }
              ]
            },
            "minimum_total_weight": 2,
            "include_unlocatable": false
          },
          "subscriber_location_weights": {
            "query_kind": "location_visits",
            "locations": [
              {
                "query_kind": "daily_location",
                "date": "2016-01-05",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              },
              {
                "query_kind": "daily_location",
                "date": "2016-01-06",
                "aggregation_unit": "admin3",
                "method": "last",
                "subscriber_subset": null
              }
            ]
          },
          "weight_threshold": 2
        }
      ],
      "stay_length_threshold": 2
    },
    "join_type": "full outer"
  }
}
